File: ruby/lib/xi/ml/tools/path_generator.rb
# encoding: utf-8



# Class generating the paths for local files/folders
class Xi::ML::Tools::PathGenerator < Xi::ML::Tools::Component
  attr_reader :paths, :res, :classes, :subsets, :preproc, :trans, :classifs

  # Initialize all the necessary paths for data files and models
  #
  # @param res [String] the base ressources path
  # @param classes [Array] the array of categories names
  # @param subsets [Array] the array of subsets names
  # @param preproc [Array] the array of preprocessing names
  # @param trans [Array] the array of transformation names
  # @param classifs [Array] the array of classifier names
  def initialize(res='', classes=[], subsets=[], preproc=[],
    trans=[], classifs=[])

    super()
    @logger.info('Generate all the necessary paths for data files and models')

    @res = res
    @classes = classes.clone
    @subsets = subsets.clone
    @preproc = preproc.clone
    @trans = trans.clone
    @classifs = classifs.clone

    @paths = {}

    # data files
    @paths['data'] = generate_data_files()

    @logger.info("The generated paths:\n#{PP.pp(@paths, '')}")
  end

  # Display the object's contents
  def to_s
    @paths
  end


  # Return the list of preprocessed training data files for each category
  # Files should already exist
  def get_train_preprocessed_files(ptype)
    files = []

    @classes.each do |category|
      filename = @paths['data'][category][:preprocessed][ptype]['train']
      Xi::ML::Tools::Utils.check_file_readable!(filename)
      files << filename
    end

    raise Xi::ML::Error::DataError, 'Empty list of files' if files.size == 0
    files
  end

  # Return the preprocessed file for given category, subset, preprocessing
  # File should already exist
  def get_preprocessed_file(category, subset, preprocess)
    begin
      filename = @paths['data'][category]['preprocessed'][preprocess][subset]

    rescue => e
      @logger.warn("Hash path #{category}/preprocessed/#{preprocess}/#{subset}"\
        + " does not exist: #{e.message}")
    end

    Xi::ML::Tools::Utils.check_file_readable!(filename)
    filename
  end

  # Return the transformed file for given category, subset,
  # transformation, preprocessing.
  # File should be created
  def get_transformed_file(category, subset, transformation, preprocess)
    format = "#{transformation}_#{preprocess}"

    begin
      filename = @paths['data'][category]['transformed'][format][subset]
    rescue => e
      @logger.warn("Hash path #{category}/transformed/#{format}/#{subset}"\
        + " does not exist: #{e.message}")
    end

    Xi::ML::Tools::Utils.create_path(filename)
    filename
  end

  # Return the classified file for given category, subset,
  # classification, transformation, preprocessing.
  # Files should be created
  def get_classified_file(category, subset,
    classification, transformation, preprocess)

    format = "#{classification}_#{transformation}_#{preprocess}"

    begin
      filename = @paths['data'][category]['classified'][format][subset]
    rescue => e
      @logger.warn("Hash path #{category}/classified/#{format}/#{subset}"\
        + " does not exist: #{e.message}")
    end

    Xi::ML::Tools::Utils.create_path(filename)
    filename
  end

  # Return the list of 'test' classified files for given
  # classification, transformation, preprocessing.
  # Files should already exist
  def get_test_classified_files(classification, transformation, preprocess)
    files = []
    @classes.each do |category|
      filename = get_classified_file(category, 'test',
        classification, transformation, preprocess)

      Xi::ML::Tools::Utils.check_file_readable!(filename)
      files << filename
    end

    raise Xi::ML::Error::DataError, 'Empty list of files' if files.size == 0
    files
  end

  # Return the file used to store the classification statistics,
  # given the classification, transformation, preprocessing names.
  # File should be created
  def get_stats_file(classification, transformation, preprocess)
    format = "#{classification}_#{transformation}_#{preprocess}"

    filename = File.join(@res, 'stats', format + '.json')
    Xi::ML::Tools::Utils.create_path(filename)

    filename
  end

  #=======================
  # 'private' methods
  #=======================

  # Generate all the data paths for each category, subset, preproc, trans, ...
  def generate_data_files
    files = {}

    # extracted data
    @classes.each do |category|
      files[category] = {}
      folder = File.join(@res, 'data', category, 'extracted')

      files[category]['extracted'] = File.join(folder, "#{category}.json")
    end

    # divided data
    @classes.each do |category|
      files[category]['divided'] = {}
      folder = File.join(@res, 'data', category, 'divided')

      @subsets.each do |subset|
        files[category]['divided'][subset] = File.join(folder,
          "#{category}_#{subset}.json")
      end
    end

    # preprocessed data
    @classes.each do |category|
      files[category]['preprocessed'] = {}

      @preproc.each do |preprocess|
        folder = File.join(@res, 'data', category, 'preprocessed', preprocess)

        files[category]['preprocessed'][preprocess] = {}

        @subsets.each do |subset|
          files[category]['preprocessed'][preprocess][subset] = File.join(
            folder, "#{category}_#{subset}.json")
        end
      end
    end

    # transformed data
    if @trans.size > 0
      @classes.each do |category|
        files[category]['transformed'] = {}

        @trans.each do |transformation|
          @preproc.each do |preprocess|
            ctrans = "#{transformation}_#{preprocess}"

            folder = File.join(@res, 'data', category, 'transformed', ctrans)
            files[category]['transformed'][ctrans] = {}

            @subsets.each do |subset|
              files[category]['transformed'][ctrans][subset] = File.join(
                folder, "#{category}_#{subset}.json")
            end
          end
        end
      end
    end

    # classified data
    if @classifs.size > 0
      @classes.each do |category|
        files[category]['classified'] = {}

        @classifs.each do |classifier|
          @trans.each do |transformation|
            @preproc.each do |preprocess|
              ctrans = "#{classifier}_#{transformation}_#{preprocess}"

              folder = File.join(@res, 'data', category, 'classified', ctrans)
              files[category]['classified'][ctrans] = {}

              @subsets.each do |subset|
                files[category]['classified'][ctrans][subset] = File.join(
                  folder, "#{category}_#{subset}.json")
              end
            end
          end
        end
      end
    end
    files
  end

  private :generate_data_files
end

File: python/lib/xi/ml/tools/path_generator.py
# -*-coding:utf-8 -*


import os
from xi.ml.tools import utils


class PathGenerator:
    """Class generating the paths for local files/folders"""

    def __init__(self, res, classes, subsets, preproc, trans, classif):
        """Initialize all the necessary paths for data files and models"""

        self.res = res
        self.classes = tuple(classes)
        self.subsets = tuple(subsets)
        self.preproc = tuple(preproc)
        self.trans = tuple(trans)
        self.classif = tuple(classif)

        self.paths = {
            'data': self._generate_data_files(),
            'models': {
                'trans': self._generate_transmodels_folders(),
                'trans_topics': self._generate_transtopics_files(),
                'classif': self._generate_classifmodels_folders()
            },
            'dictionary': self._generate_dict_files(),
            'stats': self._generate_stats_folders()
        }

    def __str__(self):
        """Display the object's content"""
        return str(self.paths)

    def preprocessed_file(self, category, subset, ptype):
        """
        Return the preprocessed data files
        for given category, subset and preprocessing.
        File should already exist.
        """

        filename = self.paths['data'][category]['preprocessed'][ptype][subset]
        utils.check_file_readable(filename)
        return filename

    def transformed_file(self, category, subset, ttype, ptype):
        """
        Return the transformed data file
        for given category, subset, transformation and preprocessing.
        """

        ctrans = "{}_{}".format(ttype, ptype)
        filename = self.paths['data'][category]['transformed'][ctrans][subset]

        utils.create_path(filename)
        return filename

    def classified_file(
            self, category, subset,
            cname, ctype, traintype, csize, ttype, ptype):

        """
        Return the classified data file
        for given category, subset, classifier, training type, data size,
        transformation and preprocessing.
        """

        if csize == -1:
            csize = 'max'

        ctrans = "{}_{}".format(ttype, ptype)
        classif = "{}_{}_{}_{}".format(cname, ctype, traintype, str(csize))
        folder = self.paths['data'][category]['classified'][ctrans]
        filename = os.path.join(
            folder, classif, "{}_{}.json".format(category, subset))

        utils.create_path(filename)
        return filename

    def stats_file(self, cname, ctype, traintype, csize, ttype, ptype):
        """
        Return the stats file
        for given classifier, training type, data size,
        transformation and preprocessing.
        """

        if csize == -1:
            csize = 'max'

        ctrans = "{}_{}".format(ttype, ptype)
        classif = "{}_{}".format(cname, ctype)
        folder = self.paths['stats'][classif][ctrans]

        classif_file = "{}_{}.json".format(traintype, str(csize))
        filename = os.path.join(folder, classif_file)

        utils.create_path(filename)
        return filename

    def preprocessed_files(self, ptype, subset):
        """
        Return the list of preprocessed 'subset' data files for each category.
        Files should already exist.
        """

        files = []

        for category in self.classes:
            filename = self.preprocessed_file(category, subset, ptype)
            files.append(filename)

        if not files:
            exit("Empty list of preprocessed files")

        return files

    def transformed_files(self, ttype, ptype, subset):
        """
        Return the list of transformed 'subset' data files for each category.
        Files should already exist.
        """

        files = []

        for category in self.classes:
            filename = self.transformed_file(category, subset, ttype, ptype)

            utils.check_file_readable(filename)
            files.append(filename)

        if not files:
            exit("Empty list of transformed files")

        return files

    def classified_files(
            self, cname, ctype, traintype, csize, ttype, ptype, subset):

        """
        Return the list of classified 'subset' data files for each category.
        Files should already exist.
        """

        files = []

        for category in self.classes:
            filename = self.classified_file(
                category, subset, cname, ctype, traintype, csize, ttype, ptype)

            utils.check_file_readable(filename)
            files.append(filename)

        if not files:
            exit("Empty list of classified files")

        return files

    def dictionary(self, ptype):
        """
        Return the dictionary's filename given the preprocessing type.
        File should be created later on.
        """

        utils.create_path(self.paths['dictionary'][ptype])
        return self.paths['dictionary'][ptype]

    def transformation_model(self, model, ttype, ptype):
        """
        Return the model's filename given the transformation and preprocessing.
        File should be created later on.
        """

        ctrans = "{}_{}".format(ttype, ptype)
        folder = self.paths['models']['trans'][ctrans]
        filename = "model_{}.bin".format(model)

        utils.create_folder(folder)
        return os.path.join(folder, filename)

    def transformation_topics(self, ttype, ptype):
        """
        Return the model topics filename
        given the transformation and preprocessing.
        File should be created later on.
        """

        ctrans = "{}_{}".format(ttype, ptype)
        topics_file = self.paths['models']['trans_topics'][ctrans]

        utils.create_path(topics_file)
        return topics_file

    def classification_model(self, ttype, ptype, cname, ctype, train, csize):
        """
        Return the model's filename given the transformation and preprocessing.
        File should be created later on.
        """

        if csize == -1:
            csize = 'max'

        ctrans = "{}_{}".format(ttype, ptype)
        cclassif = "{}_{}".format(cname, ctype)
        folder = self.paths['models']['classif'][cclassif][ctrans]
        filename = "{}_{}.bin".format(train, str(csize))

        utils.create_path(folder)
        return os.path.join(folder, filename)

    def _generate_data_files(self):
        """
        Generate all the data paths for each category, subset, preproc, trans, ...
        """

        files = {}

        # inits
        for category in self.classes:
            files[category] = {}
            files[category]['divided'] = {}
            files[category]['preprocessed'] = {}
            files[category]['transformed'] = {}
            files[category]['classified'] = {}

            # extracted data
            folder = os.path.join(self.res, 'data', category, 'extracted')
            file = "{}.json".format(category)

            utils.create_folder(folder)
            files[category]['extracted'] = os.path.join(folder, file)

            # divided data
            folder = os.path.join(self.res, 'data', category, 'divided')
            utils.create_folder(folder)

            for subset in self.subsets:
                file = "{}_{}.json".format(category, subset)
                files[category]['divided'][subset] = os.path.join(folder, file)

            # preprocessed data
            for preprocess in self.preproc:
                folder = os.path.join(
                    self.res, 'data', category, 'preprocessed', preprocess)
                utils.create_folder(folder)

                files[category]['preprocessed'][preprocess] = {}
                for subset in self.subsets:
                    file = "{}_{}.json".format(category, subset)
                    files[category]['preprocessed'][preprocess][subset] = \
                        os.path.join(folder, file)

            # transformed data
            for transformation in self.trans:
                for preprocess in self.preproc:
                    ctrans = "{}_{}".format(transformation, preprocess)

                    folder = os.path.join(
                        self.res, 'data', category, 'transformed', ctrans)
                    utils.create_folder(folder)

                    files[category]['transformed'][ctrans] = {}
                    for subset in self.subsets:
                        file = "{}_{}.json".format(category, subset)
                        files[category]['transformed'][ctrans][subset] = \
                            os.path.join(folder, file)

            # classified data
            for transformation in self.trans:
                for preprocess in self.preproc:
                    ctrans = "{}_{}".format(transformation, preprocess)

                    folder = os.path.join(
                        self.res, 'data', category, 'p_classified', ctrans)
                    utils.create_folder(folder)
                    files[category]['classified'][ctrans] = folder

        return files

    def _generate_dict_files(self):
        """
        Generate all the dictionaries paths for each preprocessing type
        """

        files = {}

        folder = os.path.join(self.res, 'dictionary')
        utils.create_folder(folder)

        for preprocess in self.preproc:
            file = "dictionary_{}.bin".format(preprocess)
            files[preprocess] = os.path.join(folder, file)

        return files

    def _generate_transmodels_folders(self):
        """
        Generate the transformation models paths for each preproc/trans type
        """

        folders = {}

        for transformation in self.trans:
            for preprocess in self.preproc:
                ctrans = "{}_{}".format(transformation, preprocess)

                folder = os.path.join(
                    self.res, 'models', 'transformations', ctrans)

                utils.create_folder(folder)
                folders[ctrans] = folder

        return folders

    def _generate_transtopics_files(self):
        """
        Generate the transformation topics paths for each preproc/trans type
        """

        files = {}

        # recover topics only for LSI/LDA models
        trans = set(self.trans).intersection(set(['LSI', 'LDA']))

        for transformation in trans:
            for preprocess in self.preproc:
                ctrans = "{}_{}".format(transformation, preprocess)
                file = "topics_top50words_{}.json".format(ctrans)

                folder = os.path.join(
                    self.res, 'models', 'transformations_topics')
                utils.create_folder(folder)

                files[ctrans] = os.path.join(folder, file)

        return files

    def _generate_classifmodels_folders(self):
        """
        Generate all the classification models paths for each preproc/trans type
        """

        folders = {}

        for cname in self.classif:
            for ctype in ['multiclass', 'multilabel']:
                classifier = "{}_{}".format(cname, ctype)

                folder = os.path.join(
                    self.res, 'models', 'classification', classifier)

                folders[classifier] = {}
                for transformation in self.trans:
                    for preprocess in self.preproc:
                        ctrans = "{}_{}".format(transformation, preprocess)
                        folders[classifier][ctrans] = os.path.join(
                            folder, ctrans)
                        utils.create_folder(folders[classifier][ctrans])

        return folders

    def _generate_stats_folders(self):
        """
        Generate the statistic files paths for each preproc/trans/classif type
        """

        folders = {}

        for cname in self.classif:
            for ctype in ['multiclass', 'multilabel']:
                classifier = "{}_{}".format(cname, ctype)

                folder = os.path.join(
                    self.res, 'stats', 'classification', classifier)

                folders[classifier] = {}
                for transformation in self.trans:
                    for preprocess in self.preproc:
                        ctrans = "{}_{}".format(transformation, preprocess)
                        folders[classifier][ctrans] = os.path.join(
                            folder, ctrans)
                        utils.create_folder(folders[classifier][ctrans])

        return folders
